### DISCUSSION TITLE
Proposal for unified development tools and environment

### DISCUSSION THREAD

**Discussion by EuleMitKeule** [dbuezas/eq3btsmart#88] [General]
To improve the stability and code quality of the integration and to increase the chances of getting this merged into `core` I would propose implementing a unified development environment and choose a set of tools and conventions to be used.

### `pyproject.toml`
* We should create a `pyproject.toml` file that can be used to configure dev tools and package settings as well as dependencies

### Dependencies and Dependency Management
* There should be a common solution for dependency management
* We could use [`poetry`](https://python-poetry.org) together with a `pyproject.toml` file
* The integration should depend on the `homeassistant` package for code completion and analysis

### Formatting and Linting
* I can strongly recommend using [`ruff`](https://github.com/astral-sh/ruff)
* `ruff` provides blazingly fast linting and code analysis and unifies dev tools for **linting**, **formatting** and **import sorting** and is compatible with VS Code via https://marketplace.visualstudio.com/items?itemName=charliermarsh.ruff

### Static Type Checking
* `mypy` could be used to perform static type checking
* This increase the visibility of and reduce bugs in the code
* `mypy` also works with VS Code via https://marketplace.visualstudio.com/items?itemName=matangover.mypy

### VS Code Settings
* VS Code can then be configured to automatically recommend the extensions mentioned above and perform code formatting automatically

### Pre-Commit
* We could use [`pre-commit`](https://pre-commit.com) to enforce the rules set by the tools and our configuration before committing
* `ruff` will automatically fix issues when committing

As I frequently implement and use this kind of development environment I would be happy to set it up for this repository if you would like me to.

**Comment by dbuezas** [comment]
That sounds great!
I'll welcome that :)
